src/audio/crossover/crossover_generic.c

```diff
@@ -225,32 +225,38 @@ static void crossover_s32_default(struct comp_data *cd,
 				  int32_t num_sinks,
 				  uint32_t frames)
 {
-	struct crossover_state *state;
-	const struct audio_stream *source_stream = bsource->data;
-	struct audio_stream *sink_stream;
-	int32_t *x, *y;
-	int ch, i, j;
-	int idx;
-	int nch = audio_stream_get_channels(source_stream);
-	int32_t out[num_sinks];
+	/* Array to hold active sink streams; initialized to null */
+	struct audio_stream *sink_stream[SOF_CROSSOVER_MAX_STREAMS] = { NULL };
+	int active_sinks = 0;	/* Counter for active sink streams */
+	int ch, i, j;		/* Indices for channels, frames, sinks */
+	struct crossover_state *state;	/* Current crossover channel state */
+	const struct audio_stream *source_stream = bsource->data; /* Source stream */
+	int32_t *x, *y;		/* Source and sink frame pointers */
+	int nch = audio_stream_get_channels(source_stream); /* Channel count */
+	int32_t out[num_sinks]; /* Frame output for sinks */
+
+	/* Identify active sinks, avoid processing null sinks later */
+	for (j = 0; j < num_sinks; j++) {
+		if (bsinks[j])
+			sink_stream[active_sinks++] = bsinks[j]->data;
+	}
 
+	/* Process for each channel */
 	for (ch = 0; ch < nch; ch++) {
-		idx = ch;
-		state = &cd->state[ch];
+		state = &cd->state[ch]; /* Set current crossover state */
+		/* Iterate over frames */
 		for (i = 0; i < frames; i++) {
-			x = audio_stream_read_frag_s32(source_stream, idx);
+			/* Read source */
+			x = audio_stream_read_frag_s32(source_stream, ch + i * nch);
```

Comment:
you replaced an `idx += nch` addition with an addition and a multiplication `ch + i * nch` - subject to any compiler optimisations. Depending on those optimisations this might or might not end up generating different code, but I personally don't think this would be faster. I'd keep the original addition, personally, I prefer this form:
```
for (i = 0, idx = ch; i < frames; i++, idx += nch) {
```
but that might be a matter of taste. The original code was run-time equivalent to this, don't think your version is an improvement.